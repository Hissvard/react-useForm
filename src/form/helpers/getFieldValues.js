const getListFieldValues = items => {
  return items.reduce((acc, item) => {
    const itemValues = getFieldValues(item)
    acc.push(itemValues)
    return acc
  }, [])
}

export const getFieldValues = (fieldData) => {
  return Object.entries(fieldData).reduce((acc, [k, v]) => {
    let value
    if (v.props) {
      if (v.props.items) {
        value = getListFieldValues(v.props.items)
      } else {
        value = v.props.value
      }
    } else {
      if (v.items) {
        value = getListFieldValues(v.items)
      } else {
        value = v.value
      }
    }

    acc[k] = value
    return acc
  }, {})
}

