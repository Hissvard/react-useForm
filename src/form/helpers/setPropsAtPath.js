export const setPropsAtPath = (fieldData, props, path) => {
  let current = fieldData
  let previousBranch
  let usedProps = false
  const size = path.length
  for (let index = 0; index < size; index++) {
    let branch = path[index]
    if (!isNaN(previousBranch) && branch === 'fields') {
      index++
      branch = path[index]
    }

    if (!usedProps && branch === 'items') {
      usedProps = true
      current = current.props
    }

    if (index === size - 1) {
      current[branch] = usedProps ? props.props : props
    } else {
      previousBranch = branch
      current = current[branch]
    }
  }

  return fieldData
}
