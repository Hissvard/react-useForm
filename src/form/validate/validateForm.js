import { Map } from 'immutable'
import { getFieldValues } from '../helpers/getFieldValues'

export const validateForm = (validateFunctions, fieldData) => {
  if (!validateFunctions) return {}

  const values = getFieldValues(fieldData)
  let results = Map()
  let validators = validateFunctions
  if (!Array.isArray(validateFunctions)) { validators = [validateFunctions] }

  for (let v of validators) {
    results = results.mergeDeep(v(values))
  }
  return results.toJS()
}
