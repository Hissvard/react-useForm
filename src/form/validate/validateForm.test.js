import { validateForm } from './validateForm'
import { getFieldValues } from '../helpers/getFieldValues'

jest.mock('../helpers/getFieldValues')

let values, fieldData
beforeEach(() => {
  fieldData = {
    fieldOne: { props: 'props1' },
  }

  values = {
    value1: 'one',
    value2: 'two',
  }
  getFieldValues.mockReturnValue(values)
})

test('no validators returns empty object', () => {
  expect(validateForm(undefined, fieldData)).toEqual({})
})

test('singleValidator returns results', () => {
  const validator = jest.fn()
  validator.mockReturnValue({
    value1: 'Error1',
    value2: 'Error2',
  })

  expect(validateForm(validator, fieldData)).toMatchSnapshot()
})

test('multiple validators returns merged results', () => {
  const validator1 = jest.fn()
  const validator2 = jest.fn()

  validator1.mockReturnValue({
    value1: 'Error1',
  })
  validator2.mockReturnValue({
    value2: 'Error2',
  })

  expect(validateForm([validator1, validator2], fieldData)).toMatchSnapshot()
})

test('multiple validators deeply merges results', () => {
  const validator1 = jest.fn()
  const validator2 = jest.fn()

  validator1.mockReturnValue({
    nested: {
      deepNested: {
        value1: 'Error1',
        value1a: 'Error1a',
      },
    },
    other1: 'OtherError1',
  })
  validator2.mockReturnValue({
    nested: {
      deepNested: {
        value2: 'Error2',
      },
    },
    other2: 'OtherError2',
  })

  expect(validateForm([validator1, validator2], fieldData)).toMatchSnapshot()
})
