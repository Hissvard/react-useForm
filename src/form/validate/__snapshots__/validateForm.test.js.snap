// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`multiple validators deeply merges results 1`] = `
Object {
  "nested": Object {
    "deepNested": Object {
      "value1": "Error1",
      "value1a": "Error1a",
      "value2": "Error2",
    },
  },
  "other1": "OtherError1",
  "other2": "OtherError2",
}
`;

exports[`multiple validators returns merged results 1`] = `
Object {
  "value1": "Error1",
  "value2": "Error2",
}
`;

exports[`singleValidator returns results 1`] = `
Object {
  "value1": "Error1",
  "value2": "Error2",
}
`;
